Noc description support and modelling in VPR

Srivat97 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Extended the FPGA architecture description to now include NoC specific information. Then updated VPR to process the NoC information provided by the user and create an internal model. Finally, the graphics was updated to display the NoC to the user.

#### Description
<!--- Describe your changes in detail -->
libarchfpga
- Updated the xml parser code to now include the additional tags used to describe the NoC. This information is then stored internally.

libvpr
- Added a new class called NocRouter which represents a router within the NoC
- Added a new class called NocLink which represents a link within the NoC (connections between routers)
- Added a new class called NocStorage which represents the entire NoC (list of routers and links that together make up the NoC)
- Created a new context called noc_ctx, this contains all the NoC related material so that it can be accessed globally
- Added a new function called setup_noc that is called when the device is being created. This is responsible for building the NoC based on the information provided in the FPGA architecture description file. This includes creating all the routers and links.
- Added a new set of command line options for the NoC.
- Added a new button to display the NoC (draws the routers and links). Also can draw the NoC link usage as a heat map.

#### Related Issue
<!--- Pull requests should be related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
The eventual goal is to have a CAD tool that is capable of targeting FPGA architectures with an embedded NoC. But first we need to be able to get specifications about the NoC, such as its topology, location on the FPGA, and any design constraints (max link bandwidth, link latency, router latency and etc..). Once we have the NoC specifications, we need to model it so that we can ensure any steps taken in the CAD flow always meet the NoC design specifications. So the initial steps of getting the NoC information from the user and modelling it as done in this pull request.


#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Unit tests were added to verify the newly developed functions
- Additionally, sample FPGA architecture description files were created that included a description of an embedded NoC. VPR was run on the sample FPGA architecture description files with arbitrary designs and the NoC was displayed visually to ensure that the new feature worked correctly.  

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

vaughnbetz commented:
Should add an architecture file to test the overall flow works (augments the unit tests).  It's OK if it's a simple architecture-in-progress.

vaughnbetz commented on vpr/src/base/noc_link.cpp:
```diff
@@ -0,0 +1,55 @@
+/*
```
Suggest making this doxygen compatible. Then you can make a NoC subcategory in the VPR API documentation on the VTR website.

Srivat97 replied:
fixed

vaughnbetz commented on vpr/src/base/noc_link.cpp:
```diff
@@ -0,0 +1,55 @@
+    return bandwidth_usage;
+}
+
+int NocLink::get_number_of_connections(void) const {
```
Would get_num_traffic_flows be a better name?
get_num_ is a bit shorter and probably just as clear.

Srivat97 replied:
fixed

vaughnbetz commented on vpr/src/base/noc_link.h:
```diff
@@ -0,0 +1,45 @@
+
+/*
+ * The NocLink class description header file.
+ *
```
This is where I meant Doxygen!

Srivat97 replied:
fixed

vaughnbetz commented on vpr/src/base/noc_link.h:
```diff
@@ -0,0 +1,45 @@
+
+    // represents the number of routed communication paths between routers that use
+    // this link Congestion is proportional to this variable.
+    int number_of_connections;
```
num_traffic_flows won't be enough --> will need currentTraffic and maybe back ids to the traffic flows that are going through here.

vaughnbetz commented on vpr/src/base/noc_router.cpp:
```diff
@@ -0,0 +1,57 @@
+/*
+ * The NocRouter defines a router in the NoC.
```
Might want to say "physical router"

Srivat97 replied:
Fixed

vaughnbetz commented on vpr/src/base/noc_storage.cpp:
```diff
@@ -0,0 +1,187 @@
+ * The NocStorage contains the following information:
+ * - A list of all the routers in the NoC (NocRouter objects)
+ * - A list of all the links in the NoC (NocLink objects)
+ * - A connections list for each router. This is simply a list
```
Same idea; most description should go in .h file.

Srivat97 replied:
Fixed

vaughnbetz commented on vpr/src/base/noc_storage.cpp:
```diff
@@ -0,0 +1,187 @@
+    return link_storage[id].get_sink_router();
+}
+
+double NocStorage::get_noc_link_bandwidth_usage(NocLinkId id) const {
```
Suggest returning references or const references (your call) to the noc_links and noc_routers so the end caller can just call their functions directly. Otherwise you have to wrap each method in the lower level object with one here in this composite object, which is likely to get tedious.

Srivat97 replied:
Fixed

vaughnbetz reviewed (changes requested):
Good overall, but some changes requested.
Most important: unit tests must test the real code (not a copy).

Srivat97 commented on libs/libarchfpga/src/arch_util.cpp:
```diff
@@ -204,6 +204,10 @@ void free_arch(t_arch* arch) {
         vtr::free(arch->clocks->clock_inf);
     }
+
+    if (arch->noc) {
```
fixed

Srivat97 commented on libs/libarchfpga/src/physical_types.h:
```diff
@@ -1773,6 +1773,29 @@ struct t_lut_element {
 
+/* Network-on-chip(NoC) Router data type used to identify 
+ * connections regarding individual routers in the network. */
+struct t_router {
```
done

Srivat97 commented on libs/libarchfpga/src/physical_types.h:
```diff
@@ -1773,6 +1773,29 @@ struct t_lut_element {
+
+/* Network-on-chip(NoC) data type used to store the network 
+ * properties and used when builidng a dedicated on-chip network*/
+struct t_noc_inf {
```
done

Srivat97 commented on libs/libarchfpga/src/physical_types.h:
```diff
@@ -1773,6 +1773,29 @@ struct t_lut_element {
+
+    std::vector<t_router> router_list;
+
+    std::string noc_router_tile_name;
```
done

Srivat97 commented on libs/libarchfpga/src/read_xml_arch_file.cpp:
```diff
@@ -4513,6 +4537,228 @@ static void ProcessClocks(pugi::xml_node Parent, t_clock_arch* clocks, const pug
+    pugiutil::expect_only_attributes(topology_tag, {}, loc_data);
+
+    // stores router information that includes the number of connections a router has within a given topology and also the number of times a router was declared in the arch file using the <router> tag
+    // in the datastructure below, the router id is the key and stored data is a pair, where the first element is describes the number router declarations and the second element describes the numbe router connections
```
fixed

Srivat97 commented on vpr/src/base/noc_router.cpp:
```diff
@@ -0,0 +1,57 @@
+ * - The router id. This represents the unique ID given by the
+ * user in the architecture description file when describing a
+ * router. THe purpose of this is to help the user identify the
+ * router when loggin information or displaying errors.
```
Fixed and moved comments to header file

Srivat97 commented on vpr/src/base/noc_router.cpp:
```diff
@@ -0,0 +1,57 @@
+}
+
+// setters
+void NocRouter::set_router_design_module_ref(std::string design_module_ref) {
```
Fixed

Srivat97 commented on vpr/src/base/noc_router.h:
```diff
@@ -0,0 +1,43 @@
+/*
+ * The NocRouter class description header file.
+ *
+ * The NocRouter class describes a router in the NoC. This file contains the definition of what a NocRouter object is. This includes the properties
```
Fixed

Srivat97 commented on vpr/src/base/noc_router.h:
```diff
@@ -0,0 +1,43 @@
+
+    // atom id and clustering block id can be used to identify which router modules the current router represents
+    // atom id is faster than string
+    std::string router_design_module_ref;
```
Fixed

Srivat97 commented on vpr/src/base/noc_storage.cpp:
```diff
@@ -0,0 +1,187 @@
+}
+
+/*
+ * The router IDs provided by the user cannot be used to
```
Changed the router "id" variable to "user_id"

Srivat97 commented on vpr/src/base/noc_storage.cpp:
```diff
@@ -0,0 +1,187 @@
+
+/*
+ * Two links are considered parallel when the source router of one link
+ * is the destination router of the second line. And when the desitnantion
```
fixed

Srivat97 commented on vpr/src/base/setup_noc.cpp:
```diff
@@ -0,0 +1,383 @@
+}
+
+/**
+ * @brief Go through the list of logical routers (routers described by the user
```
changed

Srivat97 commented on vpr/src/base/vpr_context.h:
```diff
@@ -381,6 +382,42 @@ struct FloorplanningContext : public Context {
+    /**
+     * @brief Represents the maximum allowed bandwidth for the links in the NoC (in Gbps)
+     */
+    double noc_link_bandwidth;
```
This makes sense. Moved to the NocStorage class

Srivat97 commented on vpr/test/test_setup_noc.cpp:
```diff
@@ -0,0 +1,1139 @@
+
+/*Re-defining static functions in setup_noc.cpp that are tested here*/
+
+static void identify_and_store_noc_router_tile_positions(const DeviceGrid& device_grid, std::vector<t_noc_router_tile_position>& list_of_noc_router_tiles, std::string noc_router_tile_name) {
```
Done

Srivat97 commented on vpr/src/draw/draw.cpp:
```diff
@@ -187,13 +189,26 @@ static void set_block_text(GtkWidget* widget, gint /*response_id*/, gpointer /*d
 static void run_graphics_commands(std::string commands);
 
+// draw_noc helper functions
+static ezgl::rectangle get_noc_connection_marker_bbox(const t_logical_block_type_ptr noc_router_logical_block_type);
```
Can I change some functions that I use externally from being static?

vaughnbetz replied:
Sure.  There's nothing too special about static; just a way to hide names inside a file if they aren't needed outside.  If they are needed outside, feel free to remove the static and put their declarations in the .h file.

vaughnbetz commented:
Infrstructure issue; CI should pass once I merge Alireza's change.
Good to get this merged soon; draw will have to be updated as it was heavily refactored.

Srivat97 commented on vpr/src/draw/draw_noc.cpp:
```diff
@@ -0,0 +1,232 @@
+        return result;
+    }
+
+    // if we are here then the line has a slope, so we use the cross product to determine the angle
```
Yes, it was a dot product not a cross product. Changed and included comments.

Srivat97 commented on vpr/src/draw/draw_noc.cpp:
```diff
@@ -0,0 +1,232 @@
+#    include "vpr_error.h"
+#    include "vtr_math.h"
+
+/**
```
Done

Srivat97 commented on vpr/src/draw/draw_noc.cpp:
```diff
@@ -0,0 +1,232 @@
+                link_coords.start.x += noc_connection_marker_quarter_width;
+                link_coords.end.x += noc_connection_marker_quarter_width;
+
+                link_coords.start.y += noc_connection_marker_quarter_height;
```
Will leave as is for now. I'll think of a better solution and change later.

Srivat97 commented:
@vaughnbetz This PR should be ready to merge. Fixed all requested changes or made a comment to explain why I didn't make the change. 

I also added some sample architecture files I used to test this new feature. They can be found under "$VTR_ROOT/vtr_flow/arch/noc". There are two architecture files in there, one has a mesh topology and the other has a star topology. I also updated the README found in "$VTR_ROOT/vtr_flow/" to help new users find these files and explain their purpose.

vaughnbetz commented:
Looks good -- thanks!
